### PR TITLE
WIP: workflows for building binaries

### LINKED ISSUE #124: CI, base (binary compatibility), multi-arch discussions
cc: @robertsLando @leerob 

It is much trickier to set up CI for this repo, because:

* Binary compatibility is affected by the build base (e.g. binaries compiled on 20.04 can not run on 18.04, Ubuntu vs. EL vs. etc.).
* Node.js is a huge program, and we have to compile multiple versions. It is easy to hit limits of Github Actions. 
* We want to compile binaries for different architectures. That usually requires cross-compilation, which is more complicated. It is also possible to compile with native toolchain under emulation, but emulation is very inefficient. 
* Others...

***
For build base:

I think `rhscl/devtoolset-10-toolchain-rhel7` is a great choice. It is RedHat's certified/official GCC 10 toolchain for Enterprise Linux 7. Compiled binaries are compatible with glibc 2.17 and later. This allows the compiled binaries to run on Ubuntu 14.04+, EL7+ and most other distros. Plus, RedHat's toolchain is production-grade. 

However, RedHat do not provide certified cross toolchains, and only EL8+ supports arm64. Ubuntu does provide official cross toolchains. I think we can use Ubuntu 18.04 as base to compile binaries for architectures other than amd64. 
***
For workflow:

Each Node.js build takes approx. 1 hour.

At minimum, we have to support Active LTS and 1 maintenance LTS. Optimally, we want to support all maintained versions (Current, Active LTS, maintenance LTSs) and at least 1 recently deprecated LTS. 

There are at least three platforms (`linux`, `macos` and `windows`) and at least two architectures (`x64` and `arm64`). Additional platforms may include `linuxstatic` and `alpine`. Additional architectures may include `armv7` and `x86`. 

So, minimum case: 2 (Node.js versions) * 3 (platforms) * 2 (architectures) = 12 builds. worst case: 5 * 5 * 4 - 10 (no `armv7`/`x86` for macOS) - 15 (no `x86` for Linux platforms) = 75 builds. 

Either way, it is impossible to run full CI frequently, unless you want to get an "excessive usage" email (or bill) from Github (well, idk if they really care, though). 

I think we should use `workflow_dispatch` event, so CI is triggered manually only. We should also try to split jobs of different versions/platforms/archs, so we can choose to rebuild only the necessary ones. 
***

### LINKED ISSUE #137: Binary not found
**My Command:**
PS P:\Dieser PC\Desktop\Fs19Dummy\1Discord-Bot\AutoUpdater\player\autoupdaterv3\NyboTV> pkg . --targets=node14-win-x64

**###Console:**
> pkg@4.5.0
> Fetching base Node.js binaries to PKG_CACHE_PATH
  fetched-v14.16.0-win-x64     [                    ] 0%
> Error! 404 Not Found
  https://github.com/zeit/pkg-fetch/releases/download/v2.7/uploaded-v2.7-node-v14.16.0-win-x64
> Asset not found by direct link:
  {"tag":"v2.7","name":"uploaded-v2.7-node-v14.16.0-win-x64"}
> Not found in GitHub releases:
  {"tag":"v2.7","name":"uploaded-v2.7-node-v14.16.0-win-x64"}
> Building base binary from source:
  built-v14.16.0-win-x64
> Cloning Node.js repository from GitHub...
  git                          [                    ] 0%
> Error! Error: spawn git ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
    at onErrorNT (node:internal/child_process:480:16)
    at processTicksAndRejections (node:internal/process/task_queues:81:21)
PS P:\Dieser PC\Desktop\Fs19Dummy\1Discord-Bot\AutoUpdater\player\autoupdaterv3\NyboTV>

### PR DESCRIPTION
**Pull request by robertsLando**
This PR adds required workflows to build binaries. As a starting point I have took the workflows from @jesec fork https://github.com/jesec/pkg-fetch then I have edited the upload script to take only node latest 3 lts versions found in patches.

Fixes #124 
Fixes #137

### CODE REVIEW

**Comment by robertsLando** [comment]
@jesec I would like to know your opinion here

**Comment by robertsLando** [comment]
I also found a typo here: https://github.com/jesec/pkg-fetch/blob/master/.github/workflows/build-publish.yml#L125 `amd64 shoudl be `arm64` I think

**Comment by robertsLando** [comment]
Test run here: https://github.com/robertsLando/pkg-fetch/actions/runs/690148869

**Comment by robertsLando** [comment]
@jesec Seems the linux static are throwing errors see 

https://github.com/robertsLando/pkg-fetch/runs/2202409337?check_suite_focus=true
https://github.com/robertsLando/pkg-fetch/runs/2202409418?check_suite_focus=true

File: lib/upload.js [hunk 1]
```diff
@@ -1,55 +1,102 @@
...
+import { verify } from "./verify.js";
+import { version } from "../package.json";
+import semver from 'semver';
+
+const cloud = new Cloud({ owner: 'vercel', repo: 'pkg-fetch' });
+
+/**
+ * Returns latest 3 lts versions available in patches
+ */
+function latestLts() {
```
Comment (jesec) [inline lib/upload.js:19]: We do not need this method. 

Simply remove no longer supported patches instead.

File: lib/upload.js [hunk 2]
```diff
@@ -60,4 +107,4 @@ if (!module.parent) {
     if (!error.wasReported) log.error(error);
     process.exit(2);
   });
-}
+}
```
Comment (jesec) [inline lib/upload.js:110]: EOF.

File: license.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
 MIT License
 
+Copyright (c) 2020 Vercel, Inc.
```
Comment (jesec) [inline license.md:3]: ? We are not employees of Vercel, and thus we have no authority to attribute copyright to them. Why is this here?

File: docker/Dockerfile.linux [hunk 4]
```diff
@@ -0,0 +1,13 @@
+FROM node:buster
```
Comment (jesec) [inline docker/Dockerfile.linux:1]: Note that the binaries are not going to work with any distro older than Debian Buster. 

Per previous discussions, use RHEL toolchain instead unless you have a better idea.

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jesec) [inline .vscode/settings.json:1]: Don’t commit VS code config.

File: .github/workflows/build-publish.yml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+      - name: Build
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          build-args: |
+            TARGET_TOOLCHAIN_ARCH=aarch64
+            TARGET_NODE_ARCH=arm64
+          file: ./docker/Dockerfile.linuxstatic
+          platforms: linux/arm64
```
Comment (jesec) [inline .github/workflows/build-publish.yml:104]: This is a cross toolchain. Don’t emulate arm64. Use x86-64 for toolchain.

**Review by jesec — CHANGES_REQUESTED** [review]

File: license.md (hunk 3 above)
Comment (robertsLando) [reply license.md:3]: I think it should be there it's just missing, I'm not an employees of them but AFAIK they have acquired all pkg stuff so the copyright is their. Betteer to ask @leerob to be sure

File: lib/upload.js (hunk 1 above)
Comment (robertsLando) [reply lib/upload.js:19]: Humm I would keep them there for the ones looking for a different version, this is just for use when building and allows to keep a reference of old pathces

File: docker/Dockerfile.linux (hunk 4 above)
Comment (robertsLando) [reply docker/Dockerfile.linux:1]: I have no experience in this, could you fix it?

File: license.md (hunk 3 above)
Comment (leerob) [reply license.md:3]: We didn't acquire it - we created it! This was before my time at Vercel, but yes this should be MIT Copyrighted to the Vercel GitHub org. Also... I know 2020 was a blur but it's 2021 now 😄

File: .vscode/settings.json (hunk 5 above)
Comment (leerob) [reply .vscode/settings.json:1]: Agreed! Maybe add to gitignore

File: license.md (hunk 3 above)
Comment (jesec) [reply license.md:3]: They will do it themselves. I’d rather not do someone else’s job, especially when I have absolutely no legal authority to do so.

File: license.md [hunk 7]
```diff
@@ -1,5 +1,6 @@
...
 Copyright (c) 2017 Zeit, Inc.
```
Comment (leerob) [inline license.md:4]:
```suggestion
Copyright (c) 2021 Vercel, Inc.
```

File: .github/workflows/build-publish.yml (hunk 6 above)
Comment (robertsLando) [reply .github/workflows/build-publish.yml:104]: I have no experience in this, could you fix it?

File: license.md (hunk 7 above)
Comment (robertsLando) [reply license.md]: AFAIK when editing a licnese file the old copyright MUST be kept there. You just need to add the new copyright holder

Comment (leerob) [reply license.md]: It's the same holder - Vercel used to be ZEIT. You can see this change reflected on other repos in our org.

File: lib/upload.js (hunk 1 above)
Comment (jesec) [reply lib/upload.js:19]: No. We either support them or we don’t. And the answer is that we can’t support all old versions. 

People can check them out from Git history and support them themselves if necessary. 

Plus, such logics lead to potential inconsistency and non-determinism.

File: docker/Dockerfile.linux (hunk 4 above)
Comment (jesec) [reply docker/Dockerfile.linux:1]: Next week, as I said.

**Comment by jesec** [comment]
> @jesec Seems the linux static are throwing errors see
> 
> https://github.com/robertsLando/pkg-fetch/runs/2202409337?check_suite_focus=true
> https://github.com/robertsLando/pkg-fetch/runs/2202409418?check_suite_focus=true

Missing `npm run prepare` or `npm run babel`, I assume. The issue does not happen in my repo, so probably some changes are missed.

File: license.md (hunk 3 above)
Comment (jesec) [reply license.md]: To clarify:

I know this could be hard to understand in this context because `pkg` is such an awesome and useful project. But the fact is that copyright can’t not be unilaterally attributed to someone, and we can not sign a license for Vercel. 

To make things easier to understand, imagine that someone wrote a malware or other unwanted softwares, and then put `Copyright acme inc.` in the executables.

**Comment by robertsLando** [comment]
@jesec IMO we should also pass github credentials down to dockerfiles in order to allow workflow to send binaries as soon as possible after they finish the build

**Comment by jesec** [comment]
> @jesec IMO we should also pass github credentials down to dockerfiles in order to allow workflow to send binaries as soon as possible after they finish the build instead of using the upload artifacts action

No. IMO we should remove the upload functions from the (JS-side) codes, and implement the functions in the workflows. 

Additionally, we may want to upload the binaries manually, to sync with the release.

**Comment by robertsLando** [comment]
The problem is that expecially when building with docker you could build just 1 binary before the 6 hours limit is reached and the job cancelled, you could build 2 and for the other one have no binaries uploaded... Otherwise we should use a matrix to build just one nodejs version for time

**Comment by jesec** [comment]
> The problem is that expecially when building with docker you could build just 1 binary before the 6 hours limit is reached and the job cancelled, you could build 2 and for the other one have no binaries uploaded... Otherwise we should use a matrix to build just one nodejs version for time

Each Docker run builds one binary only. Each job only executes one Docker run only.

**Comment by robertsLando** [comment]
ATM the upload script builds all available patches that's why I was asking. In this case we could make the node version a workflow input or we could create a workflow dispatcher that triggers this workflow for each node version found in patches using this: https://github.com/marketplace/actions/workflow-dispatch

**Comment by jesec** [comment]
> ATM the upload script builds all available patches that's why I was asking. In this case we could make the node version a workflow input and we could create a workflow dispatcher that triggers this workflow for each node version found in patches using this: https://github.com/marketplace/actions/workflow-dispatch

Host Node.js version is unrelated. We would have to add a function to allow building for specific Node version only.

**Comment by robertsLando** [comment]
I know that, I thought we could get the version from the process version in that way. BTW we also could pass it as an env var

**Comment by jesec** [comment]
Closed in favor of #153.
